packages/snaps-controllers/src/snaps/endowments/caveats/generic.ts

```diff
@@ -0,0 +1,48 @@
+import type { PermissionValidatorConstraint } from '@metamask/permission-controller';
+import { rpcErrors } from '@metamask/rpc-errors';
+
+/**
+ * Create a generic permission validator that validates the presence of certain caveats.
+ *
+ * This validator only validates the types of the caveats, not the values.
+ *
+ * @param caveatsToValidate - A list of objects that represent caveats.
+ * @param caveatsToValidate.type - The string defining the caveat type.
+ * @param caveatsToValidate.optional - An optional boolean flag that defines
+ * whether the caveat is optional or not.
+ * @returns A function that validates a permission.
+ */
+export function createGenericPermissionValidator(
+  caveatsToValidate: {
+    type: string;
+    optional?: boolean;
+  }[],
+): PermissionValidatorConstraint {
+  const validCaveatTypes = new Set(
+    caveatsToValidate.map((caveat) => caveat.type),
+  );
+  const requiredCaveats = caveatsToValidate.filter(
+    (caveat) => !caveat.optional,
+  );
+
+  return function ({ caveats }) {
+    const actualCaveats = caveats ?? [];
+    const passedCaveatTypes = actualCaveats.map((caveat) => caveat.type);
+    const passedCaveatsSet = new Set(passedCaveatTypes);
+
+    if (
+      // Disallow duplicates
+      passedCaveatsSet.size !== passedCaveatTypes.length ||
+      // Disallow caveats that don't match expected types
+      !actualCaveats.every((caveat) => validCaveatTypes.has(caveat.type)) ||
+      // Fail if not all required caveats are specified
+      !requiredCaveats.every((caveat) => passedCaveatsSet.has(caveat.type))
+    ) {
+      throw rpcErrors.invalidParams({
+        message: `Expected the following caveats: ${caveatsToValidate
```

Comment:
If there are duplicate caveats, this error message could be unclear. I don't think that will happen very often though.

Reply:
In the same vein, might as well have custom error messages for each scenario.

Reply:
Yeah, that makes sense!